csireverseproxy/main.go

```diff
@@ -120,28 +125,90 @@ func (s *Server) Config() *config.ProxyConfig {
-	if err != nil {
-		return err
-	}
-	s.CertFile = filepath.Join(s.Opts.TLSCertDir, s.Opts.CertFile)
-	s.KeyFile = filepath.Join(s.Opts.TLSCertDir, s.Opts.KeyFile)
-	s.Port = proxyConfig.Port
-	proxy, err := proxy.NewProxy(*proxyConfig)
-	if err != nil {
-		return err
+	// Read the config from secret if secret provided
+	if getEnv(common.EnvReverseProxyUseSecret, "false") == "true" {
+		log.Printf("Reading config using secret")
+
+		vs := viper.New()
+		proxySecret, err := config.ReadConfigFromSecret(vs)
+		if err != nil {
+			log.Printf("Error while reading config from secret: %v\n", err)
+			return err
+		}
+
+		proxyConfig, err := config.NewProxyConfigFromSecret(proxySecret, k8sUtils)
+		if err != nil {
+			log.Printf("Error while creating proxy config from secret: %v\n", err)
+			return err
+		}
+		s.CertFile = filepath.Join(s.Opts.TLSCertDir, s.Opts.CertFile)
+		s.KeyFile = filepath.Join(s.Opts.TLSCertDir, s.Opts.KeyFile)
+
+		s.Port = proxyConfig.Port
+		proxy, err := proxy.NewProxy(*proxyConfig)
+		if err != nil {
+			log.Printf("Error while creating proxy instance from secret: %v\n", err)
+			return err
+		}
+
+		log.Infof("Setting up watcher for mounted secret")
+		s.SetupConfigWatcher(k8sUtils, vs, s.configChangeSecret)
+
+		// params config map
+		vcp := viper.New()
+		paramsFilePath := getEnv(common.EnvPowermaxConfigPath, "")
+		paramsConfig, err := config.ReadParamsConfigMapFromPath(paramsFilePath, vcp)
+		if err != nil {
+			log.Printf("Error while reading from params config map: %v\n", err)
+			return err
+		}
+		if paramsConfig.Port != "" {
+			log.Infof("Setting reverseproxy port to %s", paramsConfig.Port)
+			s.Port = paramsConfig.Port
+			proxyConfig.Port = paramsConfig.Port
+		}
+
+		log.Infof("Setting up watcher for mounted params config map")
+		s.SetupConfigWatcher(k8sUtils, vcp, s.configChangeParamsConfigMap)
+
+		s.Proxy = proxy
+		s.SetConfig(proxyConfig)
+		s.SigChan = make(chan os.Signal, 1)
+
+	} else {
+		// Read the config from config map
+		log.Printf("Reading config using config map")
+		vcm := viper.New()
+		proxyConfigMap, err := config.ReadConfig(s.Opts.ConfigFileName, s.Opts.ConfigDir, vcm)
+		if err != nil {
+			return err
+		}
+		updateRevProxyLogParams(proxyConfigMap.LogFormat, proxyConfigMap.LogLevel)
+		proxyConfig, err := config.NewProxyConfig(proxyConfigMap, k8sUtils)
+		if err != nil {
+			return err
+		}
+		s.CertFile = filepath.Join(s.Opts.TLSCertDir, s.Opts.CertFile)
+		s.KeyFile = filepath.Join(s.Opts.TLSCertDir, s.Opts.KeyFile)
+		s.Port = proxyConfig.Port
+		proxy, err := proxy.NewProxy(*proxyConfig)
+		if err != nil {
+			log.Printf("Error while creating proxy instance from config map: %v\n", err)
+			return err
+		}
+
+		log.Infof("Setting up watcher for mounted reverse proxy config map")
+
+		s.SetupConfigWatcher(k8sUtils, vcm, s.configChangeConfigMap)
+
+		s.Proxy = proxy
+		s.SetConfig(proxyConfig)
+		s.SigChan = make(chan os.Signal, 1)
 	}
```

Comment:
Consider breaking this down into two smaller functions, one to hand the case for secrets and the other for configmap. These functions tend to grow. Easier too for unit testing smaller functions. I recognize that in a couple of releases the older method will go away so decide what is best.

Reply:
I decided to leave it as is, good chunk of code will need to be removed when we remove config map support.